File: android/src/main/java/com/getcapacitor/community/admob/nativeads/NativeAdPlacementValue.java
package com.getcapacitor.community.admob.nativeads;

import org.json.JSONException;
import org.json.JSONObject;

final class NativeAdPlacementValue {

    final String feedId;
    final String slotKey;
    final int generation;
    final int rectX;
    final int rectY;
    final int rectWidth;
    final int rectHeight;
    final int clipX;
    final int clipY;
    final int clipWidth;
    final int clipHeight;

    private NativeAdPlacementValue(JSONObject value, float density) throws JSONException {
        feedId = value.getString("feedId");
        slotKey = value.getString("slotKey");
        generation = value.optInt("generation", -1);
        JSONObject rect = value.getJSONObject("rect");
        JSONObject clip = value.getJSONObject("clipRect");
        rectX = toPixels(rect.getDouble("x"), density);
        rectY = toPixels(rect.getDouble("y"), density);
        rectWidth = toPixels(rect.getDouble("width"), density);
        rectHeight = toPixels(rect.getDouble("height"), density);
        clipX = toPixels(clip.getDouble("x"), density);
        clipY = toPixels(clip.getDouble("y"), density);
        clipWidth = toPixels(clip.getDouble("width"), density);
        clipHeight = toPixels(clip.getDouble("height"), density);
    }

    static NativeAdPlacementValue parse(String expectedFeedId, JSONObject value, float density) throws JSONException {
        if (!value.optBoolean("visible", false)) return null;
        NativeAdPlacementValue placement = new NativeAdPlacementValue(value, density);
        if (!expectedFeedId.equals(placement.feedId) || !placement.hasArea()) return null;
        return placement;
    }

    private boolean hasArea() {
        return rectWidth > 0 && rectHeight > 0 && clipWidth > 0 && clipHeight > 0;
    }

    private static int toPixels(double logicalPixels, float density) {
        return Math.round((float) logicalPixels * density);
    }
}

File: android/src/main/java/com/getcapacitor/community/admob/nativeads/NativeAdPlacementValue.kt
package com.getcapacitor.community.admob.nativeads

import org.json.JSONObject
import kotlin.math.roundToInt

internal class NativeAdPlacementValue private constructor(
    val feedId: String,
    val slotKey: String,
    val generation: Int,
    val rectX: Int,
    val rectY: Int,
    val rectWidth: Int,
    val rectHeight: Int,
    val clipX: Int,
    val clipY: Int,
    val clipWidth: Int,
    val clipHeight: Int,
) {
    private fun hasArea(): Boolean = rectWidth > 0 && rectHeight > 0 && clipWidth > 0 && clipHeight > 0

    companion object {
        fun parse(expectedFeedId: String, value: JSONObject, density: Float): NativeAdPlacementValue? {
            if (!value.optBoolean("visible", false)) return null
            val rect = value.getJSONObject("rect")
            val clip = value.getJSONObject("clipRect")
            val placement = NativeAdPlacementValue(
                feedId = value.getString("feedId"),
                slotKey = value.getString("slotKey"),
                generation = value.optInt("generation", -1),
                rectX = rect.pixels("x", density),
                rectY = rect.pixels("y", density),
                rectWidth = rect.pixels("width", density),
                rectHeight = rect.pixels("height", density),
                clipX = clip.pixels("x", density),
                clipY = clip.pixels("y", density),
                clipWidth = clip.pixels("width", density),
                clipHeight = clip.pixels("height", density),
            )
            return placement.takeIf { it.feedId == expectedFeedId && it.hasArea() }
        }

        private fun JSONObject.pixels(name: String, density: Float): Int = (getDouble(name) * density).roundToInt()
    }
}
